cli/graph_util.rs

```diff
@@ -0,0 +1,400 @@
+    checked_libs: HashSet<TypeLib>,
+    maybe_types: Option<Result<(ModuleSpecifier, Range), ResolutionError>>,
+  },
+  Configuration {
+    dependencies:
+      BTreeMap<String, Result<(ModuleSpecifier, Range), ResolutionError>>,
+  },
+  Error(ModuleGraphError),
+  Redirect(ModuleSpecifier),
+}
+
+/// Composes data from potentially many `ModuleGraph`s.
+#[derive(Debug, Default)]
+pub(crate) struct GraphData {
+  modules: HashMap<ModuleSpecifier, ModuleEntry>,
+  /// Map of first known referrer locations for each module. Used to enhance
+  /// error messages.
+  referrer_map: HashMap<ModuleSpecifier, Range>,
+  configurations: HashSet<ModuleSpecifier>,
+}
+
+impl GraphData {
+  /// Store data from `graph` into `self`.
+  pub(crate) fn add_graph(&mut self, graph: &ModuleGraph, reload: bool) {
+    for synthetic_module in graph.synthetic_modules() {
+      let mut dependencies = BTreeMap::new();
+      for (specifier, resolved) in &synthetic_module.dependencies {
+        if let Some(dep_result) = resolved {
+          dependencies.insert(specifier.clone(), dep_result.clone());
+          if let Ok((specifier, referrer_range)) = dep_result {
+            let entry = self.referrer_map.entry(specifier.clone());
+            entry.or_insert_with(|| referrer_range.clone());
+          }
+        }
+      }
+      self.modules.insert(
+        synthetic_module.specifier.clone(),
+        ModuleEntry::Configuration { dependencies },
+      );
+      self
+        .configurations
+        .insert(synthetic_module.specifier.clone());
+    }
+    for (specifier, result) in graph.specifiers() {
+      if !reload && self.modules.contains_key(&specifier) {
+        continue;
+      }
+      if let Some(found) = graph.redirects.get(&specifier) {
+        let module_entry = ModuleEntry::Redirect(found.clone());
+        self.modules.insert(specifier.clone(), module_entry);
+        continue;
+      }
+      match result {
+        Ok((_, media_type)) => {
+          let module = graph.get(&specifier).unwrap();
+          let code = Arc::new(module.source.as_str().to_string());
+          let dependencies = module.dependencies.clone();
+          let module_entry = ModuleEntry::Module {
+            code,
+            dependencies,
+            media_type,
+            checked_libs: Default::default(),
+            maybe_types: module
+              .maybe_types_dependency
+              .as_ref()
+              .and_then(|(_, r)| r.clone()),
+          };
+          if let Some((_, Some(Ok((specifier, referrer_range))))) =
+            &module.maybe_types_dependency
+          {
+            let specifier = graph.redirects.get(specifier).unwrap_or(specifier);
+            let entry = self.referrer_map.entry(specifier.clone());
+            entry.or_insert_with(|| referrer_range.clone());
+          }
+          for dep in module.dependencies.values() {
+            #[allow(clippy::manual_flatten)]
```

Comment:
What's the reason for ignoring this lint?

Reply:
The fix is more difficult to understand and more lines
```diff
- #[allow(clippy::manual_flatten)]
- for resolved in [&dep.maybe_code, &dep.maybe_type] {
-   if let Some(Ok((specifier, referrer_range))) = resolved {
-     let specifier =
-       graph.redirects.get(specifier).unwrap_or(specifier);
-     let entry = self.referrer_map.entry(specifier.clone());
-     entry.or_insert_with(|| referrer_range.clone());
-   }
- }
+ for (specifier, referrer_range) in
+   [&dep.maybe_code, &dep.maybe_type]
+     .into_iter()
+     .flatten()
+     .flatten()
+ {
+   let specifier =
+     graph.redirects.get(specifier).unwrap_or(specifier);
+   let entry = self.referrer_map.entry(specifier.clone());
+   entry.or_insert_with(|| referrer_range.clone());
+ }
```
It's just not a great rule for slice literals of `Option<Result<..>>`.

Reply:
I agree with @nayeemrmn here.